### ISSUE TITLE
Alt-left/right sometimes doesn't switch through the TCOs in correct left-right order

### ISSUE THREAD

**Issue by Monospace-V** [LMMS/lmms#6225]
### Bug Summary
Alt-left and alt-right are supposed to switch to viewing the TCO on the left or right respectively. However, if the order of TCOs is changed after creation, then Alt-Left/Right switches between them left/right in the order they were initially placed in, instead of the order they are actually in. (This probably has something to do with how the absolute order of the TCOs in a file remains the same, even if the associated timestamps change.)

#### Steps to reproduce
1. Create two TCOs on a piano roll/automation clip, one after the other
2. Move the first TCO after the second
3. Open the TCO on the left and try hitting Alt-Left and Alt-Right.

#### Expected behavior
Alt-Right move to the next TCO, Alt-Left do nothing.

#### Actual behavior
Alt-Left moves to the TCO on the right, Alt-Right does nothing.

#### Screenshot
(Attached: video.)
https://user-images.githubusercontent.com/76674645/143668827-379834b6-e44d-454a-8244-8a7652f73e41.mp4
In the video, I have initially created them in order 1, 3, 2, then moved 3 after 2. Then I opened 1, and hit Alt-Right twice. It switches to 3 first, then 2. Then hitting Alt-Left twice switches to 3, then 1.
See: Attached file:
[TCO Alt-LeftRight.zip](https://github.com/LMMS/lmms/files/7611246/TCO.Alt-LeftRight.zip)

#### Affected LMMS versions

1.2.0, 1.3 (alpha)

**Comment by merrickclay** [comment]
### Reproducibility:
I was able to replicate this issue on the most recent (at the time of this comment) master commit: 13e17496dcfe46723b67a322159d8486cd92b650

### Issue Notes:
This issue is also caused when TCOs are added "out of order". For example:
1. TCO added on 1st bar
2. TCO added on 3rd bar
3. TCO added on 2nd bar

### Explanation:
The `Pattern::adjacentPatternByOffset(int offset)` function is the function that ultimately contains the logic for Alt+Left and Alt+Right key presses. It calls `Track::getTCONum(TCO tco)`, which returns the index of the TCO in the `tcoVector` and `Pattern::adjacentPatternByOffset(int offset)` returns the Pattern found in the vector at index + offset (-1 for right, +1 for left).
The TCOs are pushed into the `tcoVector` object with the `push_back()` function and are not reordered or sorted to match the order they appear in the song editor (even when moving them).

### Naive Solution:
I added a `sortTCOs()` function in Track.cpp that runs` qSort()` on the `tcoVector` object, sorting the objects by ascending `startPosition`. This function was called in `TrackContentObject::movePosition()`, which is called after the creation of a new TCO and when a TCO is repositioned by clicking and dragging. This seemed to solve the problem. The issue with this solution is it is not ideal to sort the entire vector every time a TCO was created or moved.

### Proposed Solution:
- Add a function in Track.cpp that takes a TCO as a parameter, finds the TCO in the` tcoVector`, and inserts the TCO into the correct index in the `tcoVector`, to match the ordering of the TCOs in the Song Editor. This will be done by comparing the `startPosition` variable of the TCOs. This also avoids having to sort the whole vector by only reordering the TCO that has been created or moved.
-Call said function in` movePosition()`, passing in the TCO as a parameter if the `TrackContentObject::getTrack()` is true.

### Request:
I'd like to implement the solution for this issue. Here is what I propose. Comment if you can think of a better way or if there is something that I should change.

**Comment by Spekular** [comment]
If we can benefit from it elsewhere as well, I think keeping the collection of TCO's sorted seems wise. A vector isn't ideal for this though, so I wonder if there's a more suitable type for `tcoVector` or if the downsides here are overweighed by benefits elsewhere.

**Comment by merrickclay** [comment]
The proposed solution should keep the `tcoVector` sorted, just with better time complexity. Since the `tcoVector` becomes unsorted one TCO at a time (when created or moved), we should only have to worry about moving (within the vector) the TCO that was created or moved. Unless, of course, multiple TCOs can be created or moved simultaneously and asynchronously in a way that generates an odd race condition. I'm fairly new to the LMMS project and to Qt, so I don't know if this is even something that we have to worry about.

**Comment by Monospace-V** [comment]
If we have two TCO's on the same start position, then how will that be treated?

**Comment by merrickclay** [comment]
Is it possible to have two TCOs with the same start position in the same track? It doesn't seem possible as far as I can tell in the Song Editor, but maybe I'm missing something?

**Comment by Spekular** [comment]
> The proposed solution should keep the `tcoVector` sorted, just with better time complexity.
> **Time complexity:** Naive - O(n log n) Proposed - O(n)

Yes, this is what I'm basing my reply on. Obviously I shouldn't speculate on performance w/o any benchmarks, but tracks can end up with a lot of clips, and moving or cloning them (including groups) should remain responsive. Considering e.g. a linked list could achieve constant time it's might be worth switching to or adding a container more suited to re-sorting.

**Comment by Spekular** [comment]
> Is it possible to have two TCOs with the same start position in the same track?

Easily, just create multiple clips and move one, or move or clone a group w/ appropriate spacings. It should be fine to put them in an arbitrary order, or sort by length at that point (in the case of the shortcut this would make sense to some extent).

**Comment by Monospace-V** [comment]
> Is it possible to have two TCOs with the same start position in the same track? It doesn't seem possible as far as I can tell in the Song Editor, but maybe I'm missing something?
> 
> If by some chance there are two TCOs with the same startPosition in the same track, then the method I am proposing would place them right next to each other in `tcoVector`, but I don't know how I would differentiate which is displayed before the other in the Song Editor UI. Like I said, I don't think this is possible, though.

Is possible, by moving one TCO so it's over the second:
https://user-images.githubusercontent.com/76674645/147745953-d5503cb1-93f5-46f4-9e9c-cfb589183563.mp4

Plays both at the same time. The one which was created later visually overlaps the other, I think.

**Comment by merrickclay** [comment]
Is it just me or does that sound like bad UI that needs an issue of itself? A clip that you can't see in the song editor because it is overlapped by another...

**Comment by Monospace-V** [comment]
Most other DAWs have translucency/transparency behaviour to deal with visually representing overlapping clips. The fact that LMMS hides one behind the other is a problem, but a separate one. I'll file an enhancement report for it.
But the problem is still, how does Alt-Left/Right behave with these overlapping clips?

**Comment by merrickclay** [comment]
@Spekular I think you've got me sold on the idea of changing the `tcoVector` to a (sorted) doubly linked list. We could have O(n) search, insert, and move operations provided each inserted object is inserted in a sorted order. The only thing keeping me from implementing it is the question in my mind about how moving/creating multiple TCOs at the same time is implemented. If the group of TCOs is created/moved synchronously on the same thread, then this implementation would work perfectly. If the creation/movement of a group of TCOs is asynchronous or multi-threaded, then it gets a little more complicated. If you know and could tell me where in the codebase the logic that handles operations performed on multiple selected objects is, I'd greatly appreciate it.

> Most other DAWs have translucency/transparency behaviour to deal with visually representing overlapping clips. The fact that LMMS hides one behind the other is a problem, but a separate one. I'll file an enhancement request for it. But the problem is still, how does Alt-Left/Right behave with these overlapping clips?

I don't think it matters too much how we handle Alt+Left/Right with overlapping clips as long as it is predictable and consistent. I think the earlier suggestion of sorting by ascending startingPoint, then ascending length is a good one. If two TCOs have the same startingPoint and length, then we'll have to decide on some way to sort them that is consistent, something like, the most-recently moved/created TCO is sorted after the other TCO.

**Comment by Spekular** [comment]
> If you know and could tell me where in the codebase the logic that handles operations performed on multiple selected objects is, I'd greatly appreciate it.

Once instance is here, there's also code for adding/removing bars that shifts all clips and IIRC allows them to get compacted by moving too far left. https://github.com/LMMS/lmms/blob/13e55101f0a4ac5cd15b172e4e38ee79dae1915c/src/gui/TrackContentObjectView.cpp#L850
